samcli/commands/pipeline/bootstrap/cli.py

```diff
@@ -0,0 +1,222 @@
+    artifacts_bucket_arn: Optional[str],
+    create_ecr_repo: bool,
+    ecr_repo_arn: Optional[str],
+    pipeline_ip_range: Optional[str],
+    confirm_changeset: bool,
+    config_file: Optional[str],
+    config_env: Optional[str],
+) -> None:
+    """
+    implementation of `sam pipeline bootstrap` command
+    """
+    if not pipeline_user_arn:
+        pipeline_user_arn = _load_saved_pipeline_user_arn()
+
+    if interactive:
+        guided_context = GuidedContext(
+            stage_name=stage_name,
+            pipeline_user_arn=pipeline_user_arn,
+            pipeline_execution_role_arn=pipeline_execution_role_arn,
+            cloudformation_execution_role_arn=cloudformation_execution_role_arn,
+            artifacts_bucket_arn=artifacts_bucket_arn,
+            create_ecr_repo=create_ecr_repo,
+            ecr_repo_arn=ecr_repo_arn,
+            pipeline_ip_range=pipeline_ip_range,
+        )
+        guided_context.run()
+        stage_name = guided_context.stage_name
+        pipeline_user_arn = guided_context.pipeline_user_arn
+        pipeline_execution_role_arn = guided_context.pipeline_execution_role_arn
+        pipeline_ip_range = guided_context.pipeline_ip_range
+        cloudformation_execution_role_arn = guided_context.cloudformation_execution_role_arn
+        artifacts_bucket_arn = guided_context.artifacts_bucket_arn
+        create_ecr_repo = guided_context.create_ecr_repo
+        ecr_repo_arn = guided_context.ecr_repo_arn
+
+    if not stage_name:
+        raise click.UsageError("Missing required parameter '--stage-name'")
+
+    stage: Stage = Stage(
+        name=stage_name,
+        aws_profile=profile,
+        aws_region=region,
+        pipeline_user_arn=pipeline_user_arn,
+        pipeline_execution_role_arn=pipeline_execution_role_arn,
+        pipeline_ip_range=pipeline_ip_range,
+        cloudformation_execution_role_arn=cloudformation_execution_role_arn,
+        artifacts_bucket_arn=artifacts_bucket_arn,
+        create_ecr_repo=create_ecr_repo,
+        ecr_repo_arn=ecr_repo_arn,
+    )
+
+    bootstrapped: bool = stage.bootstrap(confirm_changeset=confirm_changeset)
+
+    if bootstrapped:
+        stage.print_resources_summary()
+
+        try:
+            stage.save_config(
+                config_dir=PIPELINE_CONFIG_DIR, filename=PIPELINE_CONFIG_FILENAME, cmd_names=_get_command_name()
+            )
+        except Exception:
+            # Swallow saving exceptions, if any, as the resources are already bootstrapped and the ARNs are already
+            # printed out in the screen.
+            pass
+
+
+def _load_saved_pipeline_user_arn() -> Optional[str]:
+    samconfig: SamConfig = SamConfig(config_dir=PIPELINE_CONFIG_DIR, filename=PIPELINE_CONFIG_FILENAME)
+    if not samconfig.exists():
+        return None
+    config: Dict[str, str] = samconfig.get_all(cmd_names=_get_command_name(), section="parameters")
+    return config.get("pipeline_user")
+
+
+def _get_command_name() -> List[str]:
```

Comment:
why do you want to add docstring to a private method? 
The method that do the actual logic, i.e. `get_cmd_names` is already documented https://github.com/aws/aws-sam-cli/blob/d4796787e00e8a6abe8a3aef22fb090c2beb9212/samcli/cli/context.py#L190-L204

Reply:
> why do you want to add docstring to a private method?

Because it is not intuitively clear what this function will output. If I just look at it, it involves `click` and another function. Without the click context, docstring can help to understand what it does.

Reply:
Besides, `get_cmd_names()` is a confusing function itself.

Reply:
Ummmm, are you saying the function name `_get_command_name` doesn't clearly indicates that this function returns the name of the  command?

this function just combine two lines of code together instead of duplicating them every time.

Reply:
yeah.. the confusing part is, which of the following is the output?
- `["sam", "pipeline", "bootstrap"]`
- `["sam", "pipeline"]`
- `["pipeline", "bootstrap"]`
If I cannot get the answer by looking at the function code, I think it deserves a docstring (comment)

Reply:
then this should be fixed in the public method `get_cmd_names()` itself, right?